File: src/main/java/org/springframework/samples/petclinic/BindingResult.java
package org.springframework.samples.petclinic;

import jakarta.validation.Validation;
import jakarta.validation.Validator;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

public class BindingResult extends HashMap<String, List<String>> {

	public boolean hasErrors() {
		return !isEmpty();
	}

	public boolean hasFieldErrors(String fieldName) {
		List<String> errors = get(fieldName);
		return errors != null && !errors.isEmpty();
	}

	public List<String> getFieldErrors(String fieldName) {
		return getOrDefault(fieldName, new ArrayList<>());
	}

	public void addError(String propertyName, String message) {
		ArrayList<String> newMessages = new ArrayList<>();
		var messages = putIfAbsent(propertyName, newMessages);
		if (messages == null) {
			messages = newMessages;
		}
		messages.add(message);
	}

	public List<String> getFieldErrors() {
		return values().stream().flatMap(Collection::stream).toList();
	}


	private static Validator validator;
	static {
		try (var factory = Validation.buildDefaultValidatorFactory()) {
			validator = factory.getValidator();
		}
	}
	public static @NotNull BindingResult validate(Object object) {
		var violations = validator.validate(object);
		var bindingResult = new BindingResult();
		violations.forEach(it -> bindingResult.addError(it.getPropertyPath().toString(), it.getMessage()));
		return bindingResult;
	}
}

File: src/main/kotlin/org/springframework/samples/petclinic/BindingResult.kt
package org.springframework.samples.petclinic

import jakarta.validation.Validation
import jakarta.validation.Validator

class BindingResult : HashMap<String, MutableList<String>>() {
    fun hasErrors(): Boolean {
        return !isEmpty()
    }

    fun hasFieldErrors(fieldName: String): Boolean {
        val errors = get(fieldName)
        return !errors.isNullOrEmpty()
    }

    fun getFieldErrors(fieldName: String): List<String> {
        return getOrDefault(fieldName, ArrayList())
    }

    fun addError(propertyName: String, message: String) {
        val newMessages = ArrayList<String>()
        val messages = putIfAbsent(propertyName, newMessages) ?: newMessages
        messages.add(message)
    }

    val fieldErrors: MutableList<String>
        get() = values
            .flatMap { it.toList() }
            .toMutableList()


    companion object {
        private var validator: Validator = Validation.buildDefaultValidatorFactory().use { factory ->
            factory.validator
        }

        fun validate(obj: Any): BindingResult = BindingResult().apply {
            val violations = validator.validate(obj)
            violations.forEach {
                addError(it.propertyPath.toString(), it.message)
            }
        }
    }
}
